File: backend/internal/domain/zobrist.go
package domain

var zobristTable [BoardSize * BoardSize * 2]uint64
var zobristNullMoveKey uint64

func init() {
	state := uint64(0x58A2C43F5A3B7E91)
	for i := range zobristTable {
		state += 0x9E3779B97F4A7C15
		z := state
		z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9
		z = (z ^ (z >> 27)) * 0x94D049BB133111EB
		zobristTable[i] = z ^ (z >> 31)
	}
	state += 0x9E3779B97F4A7C15
	z := state
	z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9
	z = (z ^ (z >> 27)) * 0x94D049BB133111EB
	zobristNullMoveKey = z ^ (z >> 31)
}

func ZobristKey(x, y int, player Player) uint64 {
	playerIndex := 0
	if player == PlayerBlue {
		playerIndex = 1
	}
	return zobristTable[x*BoardSize*2+y*2+playerIndex]
}

func ZobristNullMove() uint64 {
	return zobristNullMoveKey
}

File: backend/src/Caro.Domain/Zobrist.cs
namespace Caro.Domain;

public static class Zobrist
{
    private const ulong Seed = 0x58A2C43F5A3B7E91;

    private static readonly ulong[] Table = new ulong[Constants.Board.Size * Constants.Board.Size * 2];
    private static readonly ulong NullMoveKey;

    static Zobrist()
    {
        ulong state = Seed;
        for (int i = 0; i < Table.Length; i++)
        {
            state += SplitMix64.GoldenGamma;
            Table[i] = SplitMix64.Mix(state);
        }
        state += SplitMix64.GoldenGamma;
        NullMoveKey = SplitMix64.Mix(state);
    }

    public static ulong ZobristKey(int x, int y, Player player)
    {
        int playerIndex = 0;
        if (player == Player.Blue)
        {
            playerIndex = 1;
        }
        return Table[x * Constants.Board.Size * 2 + y * 2 + playerIndex];
    }

    public static ulong ZobristNullMove() => NullMoveKey;
}
